### DISCUSSION TITLE
Interim_Internship_Report_Roopa

### DISCUSSION THREAD

**Discussion by roopavsm** [petermr/petermr#26] [Show and tell]

**Comment by roopavsm** [comment]
Hi! I am Roopa, a 2nd-year undergraduate student pursuing B.Sc. Mathematics, Statistics, and Computer Science in St. Francis College for Women, Hyderabad. I am interested in Data Science and Programming.

 Here is a brief about the projects I've worked on supervised by Dr. Peter Murray-Rust during my internship which is going to be for 6 months .

**Comment by roopavsm** [comment]
# Project 1: Terpene classification

A pathway is a chain of events leading to a product. It contains multiple arrows between the elements. A metabolic pathway contains biochemical molecules. 

Terpenes are largely found as constituents of essential oils. They are mostly hydro-carbons. If an image is a metabolic pathway having terpenes in them are called as terpene pathways. 

Certain [rules](https://github.com/petermr/CEVOpen/wiki/Rules-for-Classification-of-terpene-pathway) were drafted to classify images as pathways, metabolic pathways, pre-terpene pathways and terpene pathways.

## Image annotation
To assess the familiarity of rules in terpene classification, a test on  [30 terpene pathway images](https://github.com/petermr/petermr/files/7971900/Classification.of.30.fig.csv) from Alex Pico's app was conducted with the help of 8 volunteers(I was one of them) as the following:

![image](https://user-images.githubusercontent.com/91625756/161507614-332a6b80-5927-442c-b449-fdd92a7be9f8.png)
 
## Statistical analysis of the result and its uses
- The annotation can be used in creating terpene and terpene synthase pathway dictionaries.  
- It gives a detailed understanding about the clarity of the terpene concept among volunteers and requirement of change in rules.

Now, it is crucial to decide whether the rules are adequate or not for carrying out the later procedures. Hence, we calculate the relation between volunteers' answers.

In order to select the best method deriving accurate relation between the data entries, we are considering three methods. 

The annotations were analyzed using
 
a) [Correlation Coefficient](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html) - Computes pairwise correlation of columns, excluding null values.

![image](https://user-images.githubusercontent.com/91625756/161756678-940770f0-0c0a-4326-9df4-83fed3e13146.png)

b) [Confusion Matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html) - Involves calculation of true negative, false positive, false negative and true positive values for the matrix.

![image](https://user-images.githubusercontent.com/91625756/161756858-6e1937ba-2ae8-4959-aa70-05cb07d1e861.png)

c) [Cohen's Kappa](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html) - A score that expresses the level of agreement between two annotators on a classification problem.

![image](https://user-images.githubusercontent.com/91625756/161756992-74793514-befd-4203-b0ee-3e3ed659f132.png)


## Reference
Please visit [CEVOpen Github page](https://github.com/petermr/CEVOpen)

**Reply by roopavsm** [reply]
Thank you for the feedback, sir. I will make changes accordingly.

**Reply by EmanuelFaria** [reply]
Hope my comments help, Roopa.
Commenting in GitHub is kind of cumbersome, so I copied [all three projects into this Google Doc](https://docs.google.com/document/d/1leR5lyz6x69Ss0o6jH8J5fC51hFBTBmH/edit?usp=sharing&ouid=110062979113982375503&rtpof=true&sd=true) and added comment bubbles for you there.  As always, make your own choices. I'm just throwing out thoughts for you. :-)

**Reply by EmanuelFaria** [reply]
@roopavsm Forget the Google doc. I've simplified my notes here. They may be more useful. Also, you may want to check out my notes for the other team members. Maybe you'll get something out of them. :)
___

Project 1: Terpene classification
A pathway is a chain of events leading to a product. It contains multiple arrows between the elements. A metabolic pathway contains biochemical molecules. Terpenes are largely found as constituents of essential oils. They are mostly hydro-carbons. If an image is a metabolic pathway having terpenes in them are called as terpene pathways.
> ou might want to think about how you would (concisely) describe the overall project to a 9-year old, and then describe this (sub?)project here in the same way. 
> eg. Under the subheading, you may choose to include a short description of what Image annotation means or includes/excludes in this context.
> If these components work together somehow, perhaps follow that up with a description of the features/components/capabilities that are a noteworthy part of this sub-project, and/or the how the incremental steps build upon one another, if applicable. 

Image annotation
We have obtained 30 terpene pathway images from Alex Pico's app and annotated them with the help of 8 volunteers(I was one of them) as the following:
> Suggest you link to it and/or provide an explanation of what it is/does and why you used it (as opposed to something else) and why it is important.


The data further obtained helped us know if the rules given were adequate to classify the images.

Statistical analysis of the result
> How is this data used? Is it a “feature” displayed to users, or is it an internal check that the system is functioning and calibrated properly? 

> As a layman, the Correlation Coefficient, Confusion Matrix and Cohen’s Kappa are intriguing titles. If your intended audience requires it to understand what’s happening, include links to definitions and a short description of how they work together as parts of the whole to fulfill some purpose.


The annotations were analyzed using a) Correlation Coefficient

b) Confusion Matrix

c) Cohen's Kappa.

This gave an idea of which method of analysis was more effective.
Future uses
> Ah! Here we go. Perhaps you might put this as maybe even part of the other sub-heading : Statistical analysis of the result and their Uses. And then start with the uses, the “Why we’re doing what we’re doing, before the what and how we do it.


The annotation can be used in creating terpene and terpene synthase pathway dictionaries.
It gives a detailed understanding about the clarity of the terpene concept among volunteers and requirement of change in rules.
Reference
Please visit CEVOpen Github page

**Comment by roopavsm** [comment]
# Project 2: Creating Dictionaries
In the context of this project’s requirements, dictionaries are collections of words that are all related to terpene and terpene synthase elements. For example: Hemiterpenes, Isomerase, etc. 
Unlike other dictionaries, these will contain the frequency of the words as well as their synonyms.

The words are extracted from images using 'optical character recognition' tools Tesseract and EasyOCR. Then, the words are looked up in the WikiData for their element code, description, and relatable element codes. This information is stored in dictionaries.

## Collecting words from image analysis tools
By integrating an optical character recognition tool(Tesseract) in [Pyamiimage](https://github.com/petermr/pyamiimage), we extract words from the terpene pathway images.

Let's take the figure given below as an example.

![image](https://user-images.githubusercontent.com/91625756/161525057-ad293b70-3bd3-474b-82af-ed8d9cd769ec.png)
 The words obtained are:

![image](https://user-images.githubusercontent.com/91625756/161528295-872b211d-dcd6-41f8-981b-5ea853cb5e6c.png)

Dictionary:

![image](https://user-images.githubusercontent.com/91625756/161759724-31b84a6f-3982-49b8-83f7-7e15be52fb37.png)

## Cleaning data manually 
The words acquired from the tool had a few errors, so the data was manually cleaned. Advantage of manual curation is replacing the misspelt word with the right spelling whereas for mistranslated words availability of more than one word to replace creates confusion. 
For example:  If 'ydroxy-3-me' is the output, it can be replaced with either 'hydroxy-3-methylbutyrate' or 'hydroxy-3-methylglutaryl'.  
So, we have considered using the Fuzzywuzzy matching to choose the suitable word.
(https://github.com/petermr/pyamiimage/blob/main/cleaned_phrases/Phrases.csv)

## [Fuzzywuzzy matching](https://github.com/petermr/pyamiimage/blob/main/dictionaries/fuzzywuzzy.ipynb)
Fuzzywuzzy is a string matching library in Python. This process involves manually searching the spellings in the internet to compare with the corrupted word obtained in the image analysis output. Then according to the highest index score we can choose to **replace the word manually**.
For example: 

![image](https://user-images.githubusercontent.com/91625756/167392617-8e112467-b08d-4ee2-b20f-333eb77ea52d.png)
We replace 'ydroxy-3-met' with 'hydroxy-3-methylbenzene'.

Advantages:
- Most likely to give the the right word
- Simple and easy process after obtaining the word(after manual search)

Limitations:
- The word despite having high index score may not be the actual word
- Time consuming process
- Searching the words manually is compulsory 

## Code for Dictionaries
To create dictionaries, code in Python was developed.
[Dictionary code](https://github.com/petermr/pyamiimage/blob/roopa-testing/dictionaries/dict_code.py)
We also calculate the frequency of the phrases in the list.

For example:
![image](https://user-images.githubusercontent.com/91625756/161531805-d0da6502-e7c8-42dd-a95b-2011e2558528.png)
## References
- [Dictionaries folder in Pyamiimage](https://github.com/petermr/pyamiimage/tree/roopa-testing/dictionaries)
- [Pyamiimage](https://github.com/petermr/pyamiimage)

**Reply by EmanuelFaria** [reply]
Project 2: Creating Dictionaries
A dictionary is a collection of words with their meanings and some additional information.
> Would it be more precise to say something along the lines of… 

>DEFINITION:
“In the context of this project’s requirements, dictionaries are collections of words that are all related to a specific topic. For example, “eoPlants.xml” is a dictionary including plant names IBotanical and common), wikidataID…. etc.”?
>Suggest the next line provides some context for their significance.
eg. USE: 
“Our dictionaries will be used by… to… so that…”

>It might also be useful to put in sub-sub-headings like.
 PROJECT FOCUS or SELECTED TOPIC?:
Terpenes are… 
(Perhaps a VERY short description on why they are important (synthesize compounds that can’t be made synthetically?), and how there are still many more to discover and add to the terpine pathway wiki?)
And if you think appropriate, consider swapping Project 2 for Project 1…to start with the greater context of the project, within which the sub-context of the project’s focus topic? Then here you can say “terpenes and their pathways are discussed in more detail below in Project 2” or something like that — if that makes sense to you.

> CREATION METHOD(S):
1) Manual Curation. 
(If you want to talk about that at all, a sentence or two would suffice, but it’s advantages and limitations would be interesting, to me anyway).

> 2)MACHINE CREATION/CURATION:?
2a) text analysis
2b) image analysis
ETC..

The words are extracted from images using 'optical character recognition' tools Tesseract and EasyOCR. Then, the words are looked up in the WikiData for their element code, description, and relatable element codes. This information is stored in dictionaries.
We are creating dictionaries with terpene and terpene synthase elements. Unlike other dictionaries, these will contain the frequency of the words as well as their synonyms.
Collecting words from image analysis tools
> Perhaps start with a description of the problem/challenge at this step: Relevant papers on this topic usually contain images that contain curtial information that are not included in the body text. 
This visual information includes… arrows/vectors, ….
We are developing our image analysis tool to solve that problem…
Here’s how…
> (More sub-sub-headings as useful to break the information into bite-sized chunks which also serve to make the document more “scannable” by the reader. (i.e. the Reader can get a good idea of where this paper is heading by scanning the headings before getting into the body text)

By integrating an optical character recognition tool(Tesseract) in Pyamiimage, we extract words from the terpene pathway images.
Let's take the figure given below as an example.
 The words obtained are:
\
Dictionary:
> Add labels and descriptive captions to the images, if helpful.


Fuzzywuzzy matching
Fuzzywuzzy is a string matching library in Python. As the words were not completely obtained in the output, this was used to match the names of various elements.
> Would be useful to hyperlink Python and any other jargon in the paper to layman-readable definitions.
> RE: “As the words were not completely obtained in the output, this was used to match the names of various elements.”
Does this mean that if the OCR comes up with garbage (like swapping 5 for S or 0 for O) that string matching would “learn/remember/correct” that?

> Here’s key format used in writing copy for marketing materials (such as sales letters) that might be useful here:
1. Tell me what you’re going to tell me.
2. Tell me
3. Tell me what you told me.

> I’m suggesting you consider: For each major paper section (topic change within the paper), provide n short intro that:
First, describes what you’re about to show them. Then show them with examples. And finally, restate the whole thing before moving to the next topic change… “As you can see thanks to (string matching), x problem as shown in y examples resulted in z useable output.”


Fuzzywuzzy test
> RE: fuzzywuzzy example image. Does 80 also mean there is an 80% chance that the fuzzy wuzzy result is correct? I have no idea. Just checking.

Cleaning data manually
> Regarding Sub-title “Cleaning data manually”… are these sub-heads actually “steps that are completed in order, or just components being highlighted? If steps, I suggest you name them as such: Step 1: X, Step 2: y…

The words acquired from the tool had a few errors, so the data was manually cleaned. (https://github.com/petermr/pyamiimage/blob/main/cleaned_phrases/Phrases.csv)
Code for Dictionaries
> RE: “Code for Dictionaries”… would something along the lines of “Coding for Dictionary Creation” be more precise? If your document can be understood solely by reading the headings and su-headings, you’re on the right track to naming them.

To create dictionaries, code in Python was formulated.
> is code “formulated” or “developed” or something else. Peter would know for sure.
 Dictionary code This code also calculates the frequency of occurrence of the words in the list.
> RE “This code also calculates the frequency of occurrence of the words in the list”

Suggest: “The code in the example below (Fig 2.0) does x number of things: 1, 2, 3.”

For example:

**Comment by roopavsm** [comment]
## Project 3: Alpha testing image analysis tool
This involves testing a sample of images for the accuracy of their output using Tesseract and EasyOCR tools.

# Test using Tesseract
Input:

![image](https://user-images.githubusercontent.com/91625756/161533884-08f8addc-6de9-4ebe-bab6-9c4d77ca7ba0.png)

Output:

![image](https://user-images.githubusercontent.com/91625756/161534041-3e66f836-b26a-41ef-b96b-dd0c4b3fe9a7.png)
![image](https://user-images.githubusercontent.com/91625756/161534105-d18b7f3d-8855-4fda-bc22-0b4c1fbb106d.png)

# Test using EasyOCR
Input:

Using the same image.

Output:

![image](https://user-images.githubusercontent.com/91625756/161534355-8eb9739c-30e7-440e-a4da-48ec4e5ae27f.png)
![image](https://user-images.githubusercontent.com/91625756/161534414-d7b981b3-3014-4227-8d2f-0151141f46b6.png)

## Observations
- EasyOCR gives cleaner output than Tesseract.  
- It reads phrases whereas Tesseract reads words.
- Tesseract gives more errors than EasyOCR.

## Points to work on
- If EasyOCR reads Greek alphabets, this will help us in getting the exact output from the images.
- Also, subscripts need to be processed in EasyOCR. 

## Reference
[Pyamiimage](https://github.com/petermr/pyamiimage)

**Reply by EmanuelFaria** [reply]
Project 3: Alpha testing images
> Could this be called: Alpha testing image analysis (or data extraction) or something more specific? Or perhaps keep it broad but indicate that the parts will be described in Sub-headings as “Steps”? Ergo, “Steps to Alpha-Testing Image Analysis, Verification and Data Extraction”?

This involves testing a sample of images for the accuracy of their output using Tesseract and EasyOCR tools.
Test using Tesseract
Input:
> Besides the sub-sub-heads “Input” and “Output” would it be useful/relevant/practical to include the code that does that increment?

**Comment by roopavsm** [comment]
## Project 4: Corpus of images and captions from PMC
We are creating a corpus of images containing pathway diagrams, along with their captions. Then, we run `pyamiimage` over them and curate a dictionary of terpene synthases. 

To build a corpus of images and captions, we search for 'terpene synthase' in [PMC](https://www.ncbi.nlm.nih.gov/pmc/) and explore the images section beside it. We find about 1165 images after the search. This section provides information regarding the paper title, caption title and description along with the images.  As we get pathway diagrams, tables and graphs in the result, we manually pick the pathway diagrams and create a list of them. 
https://github.com/petermr/CEVOpen/blob/master/image_links.csv

Later, we created a CProject of PMC in the `roopa-testing` branch of `pyamiimage`. This contains a Ctree of PMC IDs having images and captions alongside. This curation was done manually on Github. 56 images and captions have been uploaded so far. 
https://github.com/petermr/pyamiimage/tree/roopa-testing

Points to work on:
- The caption format should be checked.
- Test images using `pyamiimage`.
- Uploading on Github should be automated.

**Reply by EmanuelFaria** [reply]
Very nice. Would it be more precise to say, "We are creating a corpus of images containing pathway diagrams, along with their captions..."?

**Reply by roopavsm** [reply]
Yes, sir. Thank you for the feedback.
